### PR TITLE
New numerical aquifer

### PR DESCRIPTION
**Pull request by GitPaean**
It is the old implementation adapted to the new interface https://github.com/OPM/opm-common/pull/2252 and https://github.com/OPM/opm-common/pull/2246. 

The results look good. There will be more refurbishing of the implementation. 

It depends on the upstream PR https://github.com/OPM/opm-common/pull/2252

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this opm-common=2522 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2522 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
It looks like the jenkins did not grab the updated master, 

https://ci.opm-project.org/job/opm-simulators-PR-builder/1811/console

**Comment by bska** [comment]
> It looks like the jenkins did not grab the updated master,

That happens every so often, especially when rebasing a PR with "many" commits.  The best work-around I've found is to close the PR, waiting a few minutes (5?) and then reopening the PR before trying to rerun the build check.

**Comment by GitPaean** [comment]
>  especially when rebasing a PR with "many" commits. 

Just FYI, it is before the rebasing. The rebasing was trying to fix the problem.  I guess jenkins might check the upstream PR, if nothing changed, it will not recompile the upstreaming modules. I did not look into details. It is just a guess based on the symptom.

**Comment by akva2** [comment]
it's not jenkins fault. the problem is that github sometimes take its time to update the merge refs used for building.

**Comment by bska** [comment]
> > especially when rebasing a PR with "many" commits.
> 
> Just FYI, it is before the rebasing. The rebasing was trying to fix the problem.

Right.  The problem, as @akva2 [mentions](https://github.com/OPM/opm-simulators/pull/3039#issuecomment-774061687), is that GitHub doesn't recognise that the commit IDs have changed as a result of the rebase which means that the CI system ends up pulling the original commits from before the rebase.

**Comment by bska** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 please

**Comment by GitPaean** [comment]
The only testing failed are the ones related to numerical aquifers, which are lacking reference data yet.

**Comment by GitPaean** [comment]
I marked this PR as ready for review although there are still small parts missing, like the restarting, but we do not have reference to check either with it.

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 opm-tests=2252 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 opm-tests=476 please

**Comment by bska** [comment]
> The only testing failed are the ones related to numerical aquifers, which are lacking reference data yet.

That's not exactly true.  There are two [parallel simulations](https://ci.opm-project.org/job/opm-simulators-PR-builder/1815/#showFailuresLink) that fail and those don't compare to an existing reference solution on disk.  Rather, they compare to a sequential simulation of the same case.

**Comment by GitPaean** [comment]
jenkins build this opm-common=2252 opm-tests=476 please

**Comment by GitPaean** [comment]
> That's not exactly true. There are two parallel simulations that fail and those don't compare to an existing reference solution on disk. Rather, they compare to a sequential simulation of the same case.

Yes. They are also related to numerical aquifers and added in this PR. 

I just relaxed the tolerance to use the tolerance for parallel testing.

File: opm/simulators/aquifers/AquiferNumerical.hpp
Comment (GitPaean) [inline opm/simulators/aquifers/AquiferNumerical.hpp:30]: For numerical aquifers, they are part of the grid. So this class only handles the collection of the information for output.

**Comment by GitPaean** [comment]
With relaxed tolerance, the parallel running tests for the numerical aquifer cases still failed. 

I will investigate them later.

**Comment by GitPaean** [comment]
The parallel running results are not very ideal. Need to investigate more. 
[ftaq-0104.pdf](https://github.com/OPM/opm-simulators/files/5946706/ftaq-0104.pdf)

**Comment by GitPaean** [comment]
jenkins build this opm-tests=476 please

**Comment by bska** [comment]
> opm-tests=476

That update is likely to be useless by now.  Since it was created, we've merged OPM/opm-common#2295 with its associate update OPM/opm-tests#482 which conflicts with OPM/opm-tests#476.  When we're happy with the current PR&mdash;I'll start reviewing it now&mdash;we'll need to generate a new set of reference solutions for these cases.

**Comment by bska** [comment]
jenkins build this please

File: ebos/equil/initstateequil.hh
Comment (bska) [inline ebos/equil/initstateequil.hh:1969]: I suppose this bit could be split out to a separate helper function that we'd run at the end of the initialisation step.  That way it'd be applicable to all equilibration methods (i.e., cell-centre based, horizontal and, when we get it, tilted).

**Review by bska — COMMENTED** [review]
This looks good to me for the most part.  I had one comment concerning the intialisation code which looked unnecessarily restrictive but otherwise I'm in favour of getting this into the master branch. 

I am concerned, however, that the parallel runs don't produce results that pass the comparison tests to the sequential runs.  I think we need to investigate that more closely before we actually do merge this.

**Comment by GitPaean** [comment]
> I am concerned, however, that the parallel runs don't produce results that pass the comparison tests to the sequential runs. I think we need to investigate that more closely before we actually do merge this.

Yes, that is the main obstacle. I am looking into this now and will keep you updated.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=485 please

**Comment by GitPaean** [comment]
Tried to rebase, while the rebase got stuck, keeping asking whether having `git add` conflicted files although `git add` has already been done. Not sure what is the problem.

File: ebos/equil/initstateequil.hh
Comment (GitPaean) [reply ebos/equil/initstateequil.hh:1951]: It is a note when developing. I can remove it now.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=485 please

**Comment by GitPaean** [comment]
The jenkins run is supposed to fail, even for the serial regression test for numerical aquifer cases. Just performed for verification. 

After that, the reviewing comments will be addressed and the reference data needs to be `update_data` again.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=493 please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

**Comment by GitPaean** [comment]
For the parallel regression failures related to `3D_2AQU_NUM`, it looks like for the aquifer cells, the initial pressure is about 0.03 bar different, which makes the results different.  Investigating it now.

**Comment by GitPaean** [comment]
It looks like std::optional is having difficulties in parallel. Some print out for the aquifer information. The last value `cell pressure` is the `init_pressure` in `NumericalAquiferCell`. 

```
 170  rank 1 
 171   aquifer id 939 cell I J K 20 2 10 cell depth 2098.53 cell pressure 0 
 172  aquifer id 919 cell I J K 20 1 10 cell depth 2098.53 cell pressure 0 
 173 
 174  rank 2 
 175   aquifer id 939 cell I J K 20 2 10 cell depth 2098.53 cell pressure 0 
 176  aquifer id 919 cell I J K 20 1 10 cell depth 2098.53 cell pressure 0 
 177 
 178  rank 3 
 179   aquifer id 939 cell I J K 20 2 10 cell depth 2098.53 cell pressure 0 
 180  aquifer id 919 cell I J K 20 1 10 cell depth 2098.53 cell pressure 0 
 181 
 182  rank 0 
 183   aquifer id 919 cell I J K 20 1 10 cell depth 2098.53 cell pressure 2.58128e+07 
 184  aquifer id 939 cell I J K 20 2 10 cell depth 2098.53 cell pressure 2.58128e+07 

```

**Comment by GitPaean** [comment]
in the parser init_pressure was missing in the `serializeOp` of `NumericalAquiferCell`.

**Comment by GitPaean** [comment]
jenkins build this opm-common=2345 opm-tests=494 please

**Comment by GitPaean** [comment]
jenkins build this opm-common=2345 opm-tests=494 please

**Comment by bska** [comment]
> opm-tests=494

That PR most likely isn't valid anymore.  We merged OPM/opm-tests#499 yesterday and the way "update_data"-based PRs **usually** behave then means that OPM/opm-tests#494 is based off of a version that is too old.  Unless you have reason to believe otherwise, I suggest we close OPM/opm-tests#494 and generate a new reference data update from this PR.

**Comment by akva2** [comment]
there are no conflicts so the data update is valid.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Not sure why the following tests failed. 
```
mpi.compareECLFiles_flow_onephase+SPE1CASE1_WATER
mpi.compareECLFiles_flow+BASE_MODEL_1
mpi.compareECLFiles_flow+MSW_MODEL_1
mpi.compareECLFiles_flow+FAULTS_MODEL_1
mpi.compareECLFiles_flow+BASE_MSW_WSEGAICD
mpi.compareECLFiles_flow+NORNE_ATW2013_B1H_RE-PERF
```

Let us trigger another jenkins test to see.

**Comment by GitPaean** [comment]
The new test results are expected. 
```
mpi.compareECLFiles_flow+3D_2AQU_NUM
mpi.compareECLFiles_flow+3D_1AQU_3CELLS
mpi.compareParallelSim_flow+3D_2AQU_NUM
mpi.compareParallelSim_flow+3D_1AQU_3CELLS
```

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

**Comment by bska** [comment]
> there are no conflicts so the data update is possibly valid.

We've had this argument before.  Yes, the reference solution data in OPM/opm-tests#494 are valid, but that PR references an older **parent** commit.  In my experience this confuses GitHub when it tries to merge the two branches that differ only in the (binary) result files.  We've seen repeatedly that this confusion leads to GitHub picking the older fileset which then does not agree with the current simulator version.  This in turn explains why the following regression tests fail.
> ```
> mpi.compareECLFiles_flow_onephase+SPE1CASE1_WATER
> mpi.compareECLFiles_flow+BASE_MODEL_1
> mpi.compareECLFiles_flow+MSW_MODEL_1
> mpi.compareECLFiles_flow+FAULTS_MODEL_1
> mpi.compareECLFiles_flow+BASE_MSW_WSEGAICD
> mpi.compareECLFiles_flow+NORNE_ATW2013_B1H_RE-PERF
> ```

On the other hand:

> The new test results are expected.
> ```
> mpi.compareECLFiles_flow+3D_2AQU_NUM
> mpi.compareECLFiles_flow+3D_1AQU_3CELLS
> mpi.compareParallelSim_flow+3D_2AQU_NUM
> mpi.compareParallelSim_flow+3D_1AQU_3CELLS
> ```

Do we **still** expect to see failures in the parallel simulation runs (`compareParallelSim_flow+<CASE>`)?

**Comment by GitPaean** [comment]
The latest jenkins results also look fine. 
```
mpi.compareParallelSim_flow+3D_2AQU_NUM
mpi.compareParallelSim_flow+3D_1AQU_3CELLS
```

**Comment by GitPaean** [comment]
The failure causes are as follows,  the well curves are identical, let me check the details 
```
	ANQR:1: Summary file
		Fails for 3 entries
		Largest absolute error: 1.3589644e-01
		Largest relative error: 6.4341853e-02
	ANQR:2: Summary file
		Fails for 4 entries
		Largest absolute error: 1.5462379e+00
		Largest relative error: 6.4532749e-02
	ANQT:1: Summary file
		Fails for 3 entries
		Largest absolute error: 1.0175896e-01
		Largest relative error: 4.5661468e-02
	ANQT:2: Summary file
		Fails for 5 entries
		Largest absolute error: 1.1580811e+00
		Largest relative error: 4.5815767e-02

```
```
2 keywords exhibit failures
	ANQR:1: Summary file
		Fails for 1 entries
		Largest absolute error: 7.0243120e-02
		Largest relative error: 2.1278669e-02
	ANQT:1: Summary file
		Fails for 1 entries
		Largest absolute error: 3.5121560e-02
		Largest relative error: 2.1278669e-02

Program finished 
```

**Comment by GitPaean** [comment]
> We've had this argument before. Yes, the reference solution data in OPM/opm-tests#494 are valid, but that PR references an older parent commit. In my experience this confuses GitHub when it tries to merge the two branches that differ only in the (binary) result files. 

That is also my suspicion. There is something unstable, which caused the earlier confusion.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

File: opm/simulators/aquifers/AquiferNumerical.hpp
Comment (GitPaean) [inline opm/simulators/aquifers/AquiferNumerical.hpp:162]: Somehow, for parallel running, the data between aquifer 1 and aquifer 2 can be polluted here. Then different process obtain different aquifer pressure values through this function, at the end, the maximum was output into the summary file. Not understanding how this happens at all.

Comment (GitPaean) [reply opm/simulators/aquifers/AquiferNumerical.hpp:162]: It can be std::unordered_map gives different orders in different process, which makes the data between different aquifers mixed up together over different processes.  Will look for a solution tomorrow.

**Comment by GitPaean** [comment]
jenkins build this opm-common=2349 opm-tests=494 please

**Comment by GitPaean** [comment]
The big relative difference happens at the early stage when the values are relatively small.  The results are very good. I will relax the rel tolerance for parallel comparison for this case 3D_2AQU_NUM.DATA. 
```
	ANQR:1: Summary file
		Fails for 3 entries
		Largest absolute error: 1.3589644e-01
		Largest relative error: 6.4341853e-02
	ANQR:2: Summary file
		Fails for 4 entries
		Largest absolute error: 1.5462379e+00
		Largest relative error: 6.4532749e-02
```

**Comment by GitPaean** [comment]
jenkins build this opm-tests=494 please

**Comment by alfbr** [comment]
> The results are very good. I will relax the rel tolerance for parallel comparison for this case 3D_2AQU_NUM.DATA.

Please hold off increasing the relative tolerance, I would like to spend a little time to see if we have a better option. Accepting relative errors at the level of 7% is unfortunate.

**Comment by GitPaean** [comment]
> Please hold off increasing the relative tolerance, I would like to spend a little time to see if we have a better option. Accepting relative errors at the level of 7% is unfortunate.

Sure. For your reference, this is the plot for one of the values on my local machine. Visually difficult to tell, while the 6% difference happens at the third value. 

anqr_2

**Comment by bska** [comment]
> this is the plot for one of the values on my local machine. Visually difficult to tell, while the 6% difference happens at the third value.

Would you mind plotting just the first 5 values, then, to allow us to see the difference in context?

**Comment by GitPaean** [comment]
> Would you mind plotting just the first 5 values, then, to allow us to see the difference in context?

Here you go. 
anqr_2_1_6


And also, ANQR is a rate, it is based on the difference of pressures. When there is small deviations on the pressure values, the difference can change relatively significantly.

**Comment by GitPaean** [comment]
And also this is the relative difference curves
anqr_2_rel_diff

**Comment by GitPaean** [comment]
The first three values to show the difference more clearly, 
anqr_2_1-3

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=501 please

**Comment by GitPaean** [comment]
With the new setup, to pass the parallel regression test for 3D_2AQU_NUM.DATA, the absolute tolerance 0.12 is used. Hopefully it is okay
```
  add_test_compare_parallel_simulation(CASENAME numerical_aquifer_3d_2aqu
                                       FILENAME 3D_2AQU_NUM
                                       SIMULATOR flow
                                       ABS_TOL ${abs_tol_parallel}
                                       ABS_TOL 0.12
                                       REL_TOL ${coarse_rel_tol_parallel}
```

**Comment by GitPaean** [comment]
jenkins build this opm-tests=501 please

**Comment by alfbr** [comment]
> With the new setup, to pass the parallel regression test for 3D_2AQU_NUM.DATA, the absolute tolerance 0.12 is used. Hopefully it is okay

Yes, this is fine. Thank you for going the extra mile! From my end this is good to go. I would like to look into changing the dataset to make it less sensitive, but let's leave that for another pull request.

File: ebos/equil/initstateequil.hh
Comment (GitPaean) [reply ebos/equil/initstateequil.hh]: I addressed it in the new commit. Please check whether it is something you want.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=501 please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
The same phenomena as before, a lot of unrelated cases failed in the Jenkin test, probably because other updates in opm-tests are merged. To save time, a `update_data` is performed at the same time to make sure only relevant cases failed in jenkins tests.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=503 please

**Comment by GitPaean** [comment]
From my side, it is ready to go in.  I tried to rebase, but it always got stuck. Not sure why. The conflict is not difficult, but git could not recognize my solution and it gets stuck.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=503 please

**Comment by alfbr** [comment]
> From my side, it is ready to go in. I tried to rebase, but it always got stuck. Not sure why. The conflict is not difficult, but git could not recognize my solution and it gets stuck.

Not sure what that means. It seems github and jenkins is happy without a rebase, so I am ready to merge this. @bska what is your preferred course of action?

**Comment by GitPaean** [comment]
> Not sure what that means.

I think rebasing is a good practice while I explained that I did not manage to do the rebasing for this PR.

**Review by bska — APPROVED** [review]
Looks good to me.  I'll merge the reference data update (OPM/opm-tests#503) and this into master.

**Comment by bska** [comment]
Updated reference solutions have been installed on the CI system.  I'll merge this now to bring the overall system back to a consistent state.

**Pull request merged**
